fdbrpc/include/fdbrpc/AsyncFileChaos.h

```diff
@@ -91,32 +94,38 @@ class AsyncFileChaos final : public IAsyncFile, public ReferenceCounted<AsyncFil
 					pdata = (char*)arena.allocate4kAlignedBuffer(length);
 					memcpy(pdata, data, length);
 					// flip a random bit in the copied buffer
-					pdata[deterministicRandom()->randomInt(0, length)] ^= (1 << deterministicRandom()->randomInt(0, 8));
+					auto corruptedPos = deterministicRandom()->randomInt(0, length);
+					pdata[corruptedPos] ^= (1 << deterministicRandom()->randomInt(0, 8));
+					// mark the block as corrupted
+					corruptedBlock = (offset + corruptedPos) / 4096;
+					TraceEvent("CorruptedBlock")
+					    .detail("Filename", file->getFilename())
+					    .detail("Block", corruptedBlock)
+					    .log();
 
 					// increment the metric for bit flips
-					auto res = g_network->global(INetwork::enChaosMetrics);
-					if (res) {
-						ChaosMetrics* chaosMetrics = static_cast<ChaosMetrics*>(res);
+					auto chaosMetricsPointer = g_network->global(INetwork::enChaosMetrics);
+					if (chaosMetricsPointer) {
+						ChaosMetrics* chaosMetrics = static_cast<ChaosMetrics*>(chaosMetricsPointer);
 						chaosMetrics->bitFlips++;
 					}
 				}
 			}
 		}
 
-		double diskDelay = getDelay();
-		if (diskDelay == 0.0) {
-			if (pdata)
-				return holdWhile(arena, file->write(pdata, length, offset));
-
-			return file->write(data, length, offset);
-		}
-
 		// Wait for diskDelay before submitting the I/O
 		// Capture file by value in case this is destroyed during the delay
 		return mapAsync<Void, std::function<Future<Void>(Void)>, Void>(
-		    delay(diskDelay), [=, file = file](Void _) -> Future<Void> {
-			    if (pdata)
-				    return holdWhile(arena, file->write(pdata, length, offset));
+		    delay(getDelay()), [=, file = file](Void _) -> Future<Void> {
+			    if (pdata) {
+				    return map(
+				        holdWhile(arena, file->write(pdata, length, offset)), [corruptedBlock, file = file](auto res) {
+					        if (g_network->isSimulated()) {
+						        g_simulator->corruptedBlocks.template emplace(file->getFilename(), corruptedBlock);
```

Comment:
What does "template emplace" do?

Reply:
it's telling the compiler that it is calling a template. I am always a bit confused about when this is necessary. In this case I had a bunch of gcc failures where it was complaining that it can't find this method. But I also had to fix other stuff around this, so I am not sure whether this is still needed (I'll remove it and check whether it now works without the template keyword)

Reply:
It seems this is luckily not needed anymore, so I removed it